### ISSUE TITLE
Add phase4

### ISSUE THREAD

**Issue by michielbdejong** [pondersource/peppol-testsuite#1]
Working on a Docker image in https://github.com/phax/phase4/issues/54.

**Comment by michielbdejong** [comment]
Doing this for c2.pondersource.net.e
```sh
docker build -t phase4 .
docker run -it --env HOSTNAME=c2.pondersource.net -v /root/tlsKeys:/tlsKeys phase4 /bin/bash

keytool -importcert -keystore /usr/local/tomcat/conf/localhost-rsa.jks -keypass changeit -storepass changeit -alias tomcat -file /tlsKeys/tomcat.cert -trustcacerts -noprompt
vim /usr/local/tomcat/conf/server.xml

**Comment by michielbdejong** [comment]
```sh
docker run -d --name phase4 --env HOSTNAME=c2.pondersource.net -v /root/tlsKeys:/tlsKeys phase4
docker exec phase4 keytool -importcert -keystore /usr/local/tomcat/conf/localhost-rsa.jks -keypass changeit -storepass changeit -alias tomcat -file /tlsKeys/tomcat.cert -trustcacerts -noprompt
docker exec phase4 sed -i "87d;94d" /usr/local/tomcat/conf/server.xml
docker restart phase4
```

ah wait, I should probably just use LetsEncrypt for the https cert, and then find where to insert the c2/c3 certs from the zipfile I got for "eDelivery AS4 conformance testing" into phase4 in some other way.

**Comment by michielbdejong** [comment]
```sh
docker build -t phase4 .
docker run -d --name=phase4 --network="host" -e "HOSTNAME=c2.pondersource.net" -v /root/tlsKeys:/tlsKeys phase4
# docker exec phase4 certbot certonly --standalone -d 
```

I can't get the env vars to work well with `docker exec`, so doing this inside a shell instead:
```sh
docker exec -it phase4 /bin/bash
root@c2:/usr/local/tomcat# certbot certonly --standalone -d c2.pondersource.net --non-interactive --agree-tos -m=letsencrypt@pondersource.com
Saving debug log to /var/log/letsencrypt/letsencrypt.log
Plugins selected: Authenticator standalone, Installer None
Obtaining a new certificate
Performing the following challenges:
http-01 challenge for c2.pondersource.net
Waiting for verification...
Cleaning up challenges

IMPORTANT NOTES:
 - Congratulations! Your certificate and chain have been saved at:
   /etc/letsencrypt/live/c2.pondersource.net/fullchain.pem
   Your key file has been saved at:
   /etc/letsencrypt/live/c2.pondersource.net/privkey.pem
   Your cert will expire on 2021-10-10. To obtain a new or tweaked
   version of this certificate in the future, simply run certbot
   again. To non-interactively renew *all* of your certificates, run
   "certbot renew"
 - Your account credentials have been saved in your Certbot
   configuration directory at /etc/letsencrypt. You should make a
   secure backup of this folder now. This configuration directory will
   also contain certificates and private keys obtained by Certbot so
   making regular backups of this folder is ideal.
 - If you like Certbot, please consider supporting our work by:

   Donating to ISRG / Let's Encrypt:   https://letsencrypt.org/donate
   Donating to EFF:                    https://eff.org/donate-le

root@c2:/usr/local/tomcat# 
```

From https://medium.com/@mashrur123/a-step-by-step-guide-to-securing-a-tomcat-server-with-letsencrypt-ssl-certificate-65cd26290b70
```sh
openssl pkcs12 -export -out /tmp/c2.pondersource.net_fullchain_and_key.p12 \
    -in /etc/letsencrypt/live/c2.pondersource.net/fullchain.pem \
    -inkey /etc/letsencrypt/live/c2.pondersource.net/privkey.pem \
    -name tomcat
keytool -importkeystore \
    -deststorepass samplePassword -destkeypass samplePassword -destkeystore /tmp/tomcat.jks \
    -srckeystore /tmp/c2.pondersource.net_fullchain_and_key.p12  -srcstoretype PKCS12 -srcstorepass changeit \
    -alias tomcat
```

Env vars to manage:

* DOMAIN
* KEYSTORES_PASS
*

**Comment by michielbdejong** [comment]
docker build -t phase4 -f docker/Dockerfile .
docker run -d --name=phase4 --network="host" -e "HOSTNAME=c2.pondersource.net" -v /root/tlsKeys:/tlsKeys phase4
docker exec phase4 /root/init.sh

**Comment by michielbdejong** [comment]
Hm, that isn't really working yet, debugging with:
```sh
docker exec -it  -e "DOMAIN=c2.pondersource.net" -e "EMAIL=letsencrypt@pondersource.com" -v /root/as4Keys:/as4Keys phase4 /bin/bash
```

**Comment by michielbdejong** [comment]
Yay, https://c2.pondersource.net/ is up :)
Now need to get the as4 cert into phase4 somehow

**Comment by michielbdejong** [comment]
-e CONFIG_FILE=/root/phase4.properties
-e STOREPASS=123456
-e KEYPASS=123456
-e KEYNAME=pondersource-c2

touch /root/phase4.properties
echo org.apache.wss4j.crypto.merlin.keystore.type="PKCS12" >> /root/phase4.properties
echo org.apache.wss4j.crypto.merlin.keystore.file="/as4Keys/as4.pks" >> /root/phase4.properties
echo org.apache.wss4j.crypto.merlin.keystore.password="$STOREPASS" >> /root/phase4.properties
echo org.apache.wss4j.crypto.merlin.keystore.alias="$KEYNAME" >> /root/phase4.properties
echo org.apache.wss4j.crypto.merlin.keystore.private.password="$KEYPASS" >> /root/phase4.properties


keytool -list -v -keystore keystore.pks

**Comment by michielbdejong** [comment]
From what I understand, both https://c2.pondersource.net/as4 and https://c3.pondersource.net/as4 should now be serving /usr/local/tomcat/webapps/phase4 but they still look like 404s...

**Comment by michielbdejong** [comment]
I could also not see https://github.com/efsavage/hello-world-war appear on my docker+tomcat so that's a good minimal steps-to-reproduce to debug that further. Will continue with that tomorrow.

**Comment by michielbdejong** [comment]
With `cp -r /usr/local/tomcat/webapps.dist/* /usr/local/tomcat/webapps/` it did work, so that's a starting point!

**Comment by michielbdejong** [comment]
Something in META-INF is stopping the as4 app from working.

**Comment by michielbdejong** [comment]
Ah!
```
[2021-07-13T08:42:58,505] [phase4-peppol] [WARN ] [Catalina-utility-1] No such key store 'keys/test-ap-2021.p12': Failed to open key store 'keys/test-ap-2021.p12' -- com.helger.security.keystore.KeyStoreHelper.loadKeyStore(KeyStoreHelper.java:243)
```

**Comment by michielbdejong** [comment]
ok, keystore location fixed, now got one unidentified listener failing to start:
```
NOTE: Picked up JDK_JAVA_OPTIONS:  --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.rmi/sun.rmi.transport=ALL-UNNAMED
NOTE: Picked up JDK_JAVA_OPTIONS:  --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.rmi/sun.rmi.transport=ALL-UNNAMED
13-Jul-2021 09:05:21.179 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Server version name:   Apache Tomcat/9.0.50
13-Jul-2021 09:05:21.186 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Server built:          Jun 28 2021 08:46:44 UTC
13-Jul-2021 09:05:21.186 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Server version number: 9.0.50.0
13-Jul-2021 09:05:21.187 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log OS Name:               Linux
13-Jul-2021 09:05:21.188 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log OS Version:            5.4.0-73-generic
13-Jul-2021 09:05:21.188 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Architecture:          amd64
13-Jul-2021 09:05:21.189 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Java Home:             /usr/local/openjdk-11
13-Jul-2021 09:05:21.192 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log JVM Version:           11.0.11+9
13-Jul-2021 09:05:21.193 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log JVM Vendor:            Oracle Corporation
13-Jul-2021 09:05:21.193 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log CATALINA_BASE:         /usr/local/tomcat
13-Jul-2021 09:05:21.194 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log CATALINA_HOME:         /usr/local/tomcat
13-Jul-2021 09:05:21.228 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: --add-opens=java.base/java.lang=ALL-UNNAMED
13-Jul-2021 09:05:21.229 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: --add-opens=java.base/java.io=ALL-UNNAMED
13-Jul-2021 09:05:21.232 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: --add-opens=java.base/java.util=ALL-UNNAMED
13-Jul-2021 09:05:21.233 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: --add-opens=java.base/java.util.concurrent=ALL-UNNAMED
13-Jul-2021 09:05:21.233 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: --add-opens=java.rmi/sun.rmi.transport=ALL-UNNAMED
13-Jul-2021 09:05:21.233 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: -Djava.util.logging.config.file=/usr/local/tomcat/conf/logging.properties
13-Jul-2021 09:05:21.234 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: -Djava.util.logging.manager=org.apache.juli.ClassLoaderLogManager
13-Jul-2021 09:05:21.234 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: -Djdk.tls.ephemeralDHKeySize=2048
13-Jul-2021 09:05:21.234 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: -Djava.protocol.handler.pkgs=org.apache.catalina.webresources
13-Jul-2021 09:05:21.235 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: -Dorg.apache.catalina.security.SecurityListener.UMASK=0027
13-Jul-2021 09:05:21.236 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: -Dignore.endorsed.dirs=
13-Jul-2021 09:05:21.237 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: -Dcatalina.base=/usr/local/tomcat
13-Jul-2021 09:05:21.237 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: -Dcatalina.home=/usr/local/tomcat
13-Jul-2021 09:05:21.238 INFO [main] org.apache.catalina.startup.VersionLoggerListener.log Command line argument: -Djava.io.tmpdir=/usr/local/tomcat/temp
13-Jul-2021 09:05:21.270 INFO [main] org.apache.catalina.core.AprLifecycleListener.lifecycleEvent Loaded Apache Tomcat Native library [1.2.30] using APR version [1.6.5].
13-Jul-2021 09:05:21.273 INFO [main] org.apache.catalina.core.AprLifecycleListener.lifecycleEvent APR capabilities: IPv6 [true], sendfile [true], accept filters [false], random [true], UDS [true].
13-Jul-2021 09:05:21.273 INFO [main] org.apache.catalina.core.AprLifecycleListener.lifecycleEvent APR/OpenSSL configuration: useAprConnector [false], useOpenSSL [true]
13-Jul-2021 09:05:21.291 INFO [main] org.apache.catalina.core.AprLifecycleListener.initializeSSL OpenSSL successfully initialized [OpenSSL 1.1.1d  10 Sep 2019]
13-Jul-2021 09:05:22.516 INFO [main] org.apache.coyote.AbstractProtocol.init Initializing ProtocolHandler ["http-nio-8080"]
13-Jul-2021 09:05:22.613 INFO [main] org.apache.coyote.http11.AbstractHttp11Protocol.configureUpgradeProtocol The ["https-openssl-apr-443"] connector has been configured to support negotiation to [h2] via ALPN
13-Jul-2021 09:05:22.618 INFO [main] org.apache.coyote.AbstractProtocol.init Initializing ProtocolHandler ["https-openssl-apr-443"]
13-Jul-2021 09:05:22.955 INFO [main] org.apache.catalina.startup.Catalina.load Server initialization in [2585] milliseconds
13-Jul-2021 09:05:23.162 INFO [main] org.apache.catalina.core.StandardService.startInternal Starting service [Catalina]
13-Jul-2021 09:05:23.164 INFO [main] org.apache.catalina.core.StandardEngine.startInternal Starting Servlet engine: [Apache Tomcat/9.0.50]
13-Jul-2021 09:05:23.223 INFO [main] org.apache.catalina.startup.HostConfig.deployWAR Deploying web application archive [/usr/local/tomcat/webapps/as4.war]
13-Jul-2021 09:05:29.554 INFO [main] org.apache.jasper.servlet.TldScanner.scanJars At least one JAR was scanned for TLDs yet contained no TLDs. Enable debug logging for this logger for a complete list of JARs that were scanned but no TLDs were found in them. Skipping unneeded JARs during scanning can improve startup time and JSP compilation time.
[2021-07-13T09:05:32,063] [phase4-peppol] [INFO ] [main] Using '/usr/local/tomcat/phase4-data' as the data path -- com.helger.photon.app.io.WebFileIO.initPaths(WebFileIO.java:92)
[2021-07-13T09:05:32,113] [phase4-peppol] [INFO ] [main] Using '/usr/local/tomcat/webapps/as4' as the servlet context path -- com.helger.photon.app.io.WebFileIO.initPaths(WebFileIO.java:99)
[2021-07-13T09:05:32,145] [phase4-peppol] [INFO ] [main] Session passivation is now disabled -- com.helger.web.scope.mgr.WebScopeManager.setSessionPassivationAllowed(WebScopeManager.java:99)
[2021-07-13T09:05:32,203] [phase4-peppol] [INFO ] [main] Default PasswordHashCreator algorithm set to 'PBKDF2_1000_48' -- com.helger.photon.security.password.hash.PasswordHashCreatorManager.setDefaultPasswordHashCreatorAlgorithm(PasswordHashCreatorManager.java:147)
[2021-07-13T09:05:32,212] [phase4-peppol] [INFO ] [main] Set global password constraints to [PasswordConstraintList@0x2ee949d4: constraints=[[PasswordConstraintMinLength@0x057e9cd2: minLength=6]]] -- com.helger.photon.security.password.GlobalPasswordSettings.setPasswordConstraintList(GlobalPasswordSettings.java:105)
[2021-07-13T09:05:32,322] [phase4-peppol] [INFO ] [main] RequestTrackerMonitor was installed successfully. -- com.helger.xservlet.requesttrack.RequestTracker.<init>(RequestTracker.java:130)
MailcapCommandMap: load HOME
new MailcapFile: file /root/.mailcap
MailcapCommandMap: load SYS
new MailcapFile: file /usr/local/openjdk-11/conf/mailcap
MailcapCommandMap: load JAR
MailcapCommandMap: getResources
MailcapCommandMap: URL jar:file:/usr/local/tomcat/webapps/as4/WEB-INF/lib/bcmail-jdk15on-1.68.jar!/META-INF/mailcap
new MailcapFile: InputStream
parse: application/pkcs7-signature;; x-java-content-handler=org.bouncycastle.mail.smime.handlers.pkcs7_signature
  Type: application/pkcs7-signature
    Command: content-handler, Class: org.bouncycastle.mail.smime.handlers.pkcs7_signature
parse: application/pkcs7-mime;; x-java-content-handler=org.bouncycastle.mail.smime.handlers.pkcs7_mime
  Type: application/pkcs7-mime
    Command: content-handler, Class: org.bouncycastle.mail.smime.handlers.pkcs7_mime
parse: application/x-pkcs7-signature;; x-java-content-handler=org.bouncycastle.mail.smime.handlers.x_pkcs7_signature
  Type: application/x-pkcs7-signature
    Command: content-handler, Class: org.bouncycastle.mail.smime.handlers.x_pkcs7_signature
parse: application/x-pkcs7-mime;; x-java-content-handler=org.bouncycastle.mail.smime.handlers.x_pkcs7_mime
  Type: application/x-pkcs7-mime
    Command: content-handler, Class: org.bouncycastle.mail.smime.handlers.x_pkcs7_mime
parse: multipart/signed;; x-java-content-handler=org.bouncycastle.mail.smime.handlers.multipart_signed
  Type: multipart/signed
    Command: content-handler, Class: org.bouncycastle.mail.smime.handlers.multipart_signed
MailcapCommandMap: successfully loaded mailcap file from URL: jar:file:/usr/local/tomcat/webapps/as4/WEB-INF/lib/bcmail-jdk15on-1.68.jar!/META-INF/mailcap
MailcapCommandMap: URL jar:file:/usr/local/tomcat/webapps/as4/WEB-INF/lib/jakarta.mail-1.6.5.jar!/META-INF/mailcap
new MailcapFile: InputStream
parse: text/plain;;		x-java-content-handler=com.sun.mail.handlers.text_plain
  Type: text/plain
    Command: content-handler, Class: com.sun.mail.handlers.text_plain
parse: text/html;;		x-java-content-handler=com.sun.mail.handlers.text_html
  Type: text/html
    Command: content-handler, Class: com.sun.mail.handlers.text_html
parse: text/xml;;		x-java-content-handler=com.sun.mail.handlers.text_xml
  Type: text/xml
    Command: content-handler, Class: com.sun.mail.handlers.text_xml
parse: multipart/*;;		x-java-content-handler=com.sun.mail.handlers.multipart_mixed; x-java-fallback-entry=true
  Type: multipart/*
    Command: content-handler, Class: com.sun.mail.handlers.multipart_mixed
parse: message/rfc822;;	x-java-content-handler=com.sun.mail.handlers.message_rfc822
  Type: message/rfc822
    Command: content-handler, Class: com.sun.mail.handlers.message_rfc822
MailcapCommandMap: successfully loaded mailcap file from URL: jar:file:/usr/local/tomcat/webapps/as4/WEB-INF/lib/jakarta.mail-1.6.5.jar!/META-INF/mailcap
MailcapCommandMap: load DEF
new MailcapFile: InputStream
parse: image/gif;;		x-java-view=com.sun.activation.viewers.ImageViewer
  Type: image/gif
    Command: view, Class: com.sun.activation.viewers.ImageViewer
parse: image/jpeg;;		x-java-view=com.sun.activation.viewers.ImageViewer
  Type: image/jpeg
    Command: view, Class: com.sun.activation.viewers.ImageViewer
parse: text/*;;		x-java-view=com.sun.activation.viewers.TextViewer
  Type: text/*
    Command: view, Class: com.sun.activation.viewers.TextViewer
parse: text/*;;		x-java-edit=com.sun.activation.viewers.TextEditor
  Type: text/*
    Command: edit, Class: com.sun.activation.viewers.TextEditor
Merging commands for type text/*
MailcapCommandMap: successfully loaded mailcap file: /META-INF/mailcap.default
MailcapCommandMap: createDataContentHandler for multipart/related
  search DB #1
  search DB #2
  search DB #3
  search fallback DB #1
  search fallback DB #2
    got content-handler
      class com.sun.mail.handlers.multipart_mixed
[2021-07-13T09:05:32,608] [phase4-peppol] [INFO ] [main] Moving audit files -- com.helger.photon.audit.AuditManager.<init>(AuditManager.java:157)
[2021-07-13T09:05:32,615] [phase4-peppol] [INFO ] [main] Finished moving 0 files -- com.helger.photon.audit.AuditManager.<init>(AuditManager.java:169)
[2021-07-13T09:05:32,939] [phase4-peppol] [INFO ] [main] PhotonSecurityManager was initialized -- com.helger.photon.security.mgr.PhotonSecurityManager.onAfterInstantiation(PhotonSecurityManager.java:215)
[2021-07-13T09:05:33,343] [phase4-peppol] [INFO ] [main] MetaAS4Manager is initialized with in-memory data structures -- com.helger.phase4.mgr.MetaAS4Manager.<clinit>(MetaAS4Manager.java:51)
[2021-07-13T09:05:33,351] [phase4-peppol] [INFO ] [main] Creating AS4 managers using factory class com.helger.phase4.mgr.ManagerFactoryInMemory -- com.helger.phase4.mgr.MetaAS4Manager.onAfterInstantiation(MetaAS4Manager.java:102)
[2021-07-13T09:05:33,408] [phase4-peppol] [INFO ] [main] 1 AS4 profile is registered  -- com.helger.phase4.profile.AS4ProfileManager._registerAll(AS4ProfileManager.java:71)
[2021-07-13T09:05:33,410] [phase4-peppol] [INFO ] [main] MetaAS4Manager was initialized -- com.helger.phase4.mgr.MetaAS4Manager.onAfterInstantiation(MetaAS4Manager.java:113)
[2021-07-13T09:05:33,636] [phase4-peppol] [INFO ] [main] Using default implementation for ThreadExecutor -- com.helger.quartz.impl.StdSchedulerFactory._instantiate(StdSchedulerFactory.java:844)
[2021-07-13T09:05:33,643] [phase4-peppol] [INFO ] [main] Job execution threads will use class loader of thread: main -- com.helger.quartz.simpl.SimpleThreadPool.initialize(SimpleThreadPool.java:258)
[2021-07-13T09:05:33,673] [phase4-peppol] [INFO ] [main] Initialized 10 worker threads -- com.helger.quartz.simpl.SimpleThreadPool.initialize(SimpleThreadPool.java:270)
[2021-07-13T09:05:33,699] [phase4-peppol] [INFO ] [main] QuartzSchedulerThread Inheriting ContextClassLoader of thread: main -- com.helger.quartz.core.QuartzSchedulerThread.<init>(QuartzSchedulerThread.java:90)
[2021-07-13T09:05:33,710] [phase4-peppol] [INFO ] [main] Initialized Scheduler Signaller of type: class com.helger.quartz.core.SchedulerSignaler -- com.helger.quartz.core.SchedulerSignaler.<init>(SchedulerSignaler.java:47)
[2021-07-13T09:05:33,712] [phase4-peppol] [INFO ] [main] Mini Quartz Scheduler v.4.2.0 created. -- com.helger.quartz.core.QuartzScheduler.<init>(QuartzScheduler.java:162)
[2021-07-13T09:05:33,714] [phase4-peppol] [INFO ] [main] JobFactory set to: com.helger.quartz.simpl.SimpleJobFactory@47166740 -- com.helger.quartz.core.QuartzScheduler.setJobFactory(QuartzScheduler.java:2201)
[2021-07-13T09:05:33,715] [phase4-peppol] [INFO ] [main] RAMJobStore initialized. -- com.helger.quartz.simpl.RAMJobStore.initialize(RAMJobStore.java:120)
[2021-07-13T09:05:33,716] [phase4-peppol] [INFO ] [main] Scheduler meta-data: Mini Quartz Scheduler (v4.2.0) 'MiniQuartzScheduler' with instanceId 'NON_CLUSTERED'
  Scheduler class: 'com.helger.quartz.core.QuartzScheduler' - running locally.
  NOT STARTED.
  Currently in standby mode.
  Number of jobs executed: 0
  Using thread pool 'com.helger.quartz.simpl.SimpleThreadPool' - with 10 threads.
  Using job-store 'com.helger.quartz.simpl.RAMJobStore' - which does not support persistence. and is not clustered.
 -- com.helger.quartz.core.QuartzScheduler.initialize(QuartzScheduler.java:171)
[2021-07-13T09:05:33,718] [phase4-peppol] [INFO ] [main] Quartz scheduler 'MiniQuartzScheduler' initialized from default resource file in Quartz package: 'quartz.properties' -- com.helger.quartz.impl.StdSchedulerFactory._instantiate(StdSchedulerFactory.java:969)
[2021-07-13T09:05:33,719] [phase4-peppol] [INFO ] [main] Quartz scheduler version: 4.2.0 -- com.helger.quartz.impl.StdSchedulerFactory._instantiate(StdSchedulerFactory.java:971)
[2021-07-13T09:05:33,726] [phase4-peppol] [INFO ] [main] Scheduler MiniQuartzScheduler_$_NON_CLUSTERED started. -- com.helger.quartz.core.QuartzScheduler.start(QuartzScheduler.java:329)
[2021-07-13T09:05:33,759] [phase4-peppol] [INFO ] [main] Succesfully scheduled job 'AS4DuplicateCleanupJob-10' with TriggerKey DEFAULT.6da64b5bd2ee-84d95c5c-c93c-4dce-8efb-50fddd78d167 - starting at 2021-07-13T09:05:33.729 -- com.helger.schedule.quartz.GlobalQuartzScheduler.scheduleJob(GlobalQuartzScheduler.java:171)
[2021-07-13T09:05:33,907] [phase4-peppol] [INFO ] [main] Successfully loaded configured key store from the crypto factory -- com.helger.phase4.peppol.server.servlet.Phase4PeppolWebAppListener._initPeppolAS4(Phase4PeppolWebAppListener.java:199)
13-Jul-2021 09:05:34.108 SEVERE [main] org.apache.catalina.core.StandardContext.startInternal One or more listeners failed to start. Full details will be found in the appropriate container log file
13-Jul-2021 09:05:34.127 SEVERE [main] org.apache.catalina.core.StandardContext.startInternal Context [/as4] startup failed due to previous errors
13-Jul-2021 09:05:36.025 INFO [main] org.apache.catalina.startup.HostConfig.deployWAR Deployment of web application archive [/usr/local/tomcat/webapps/as4.war] has finished in [12,802] ms
13-Jul-2021 09:05:36.039 INFO [main] org.apache.coyote.AbstractProtocol.start Starting ProtocolHandler ["http-nio-8080"]
13-Jul-2021 09:05:36.085 INFO [main] org.apache.coyote.AbstractProtocol.start Starting ProtocolHandler ["https-openssl-apr-443"]
13-Jul-2021 09:05:36.115 INFO [main] org.apache.catalina.startup.Catalina.start Server startup in [13156] milliseconds

**Comment by michielbdejong** [comment]
Ah,
```
root@c2:/usr/local/tomcat# cat logs/localhost.2021-07-13.log 
13-Jul-2021 09:05:34.102 SEVERE [main] org.apache.catalina.core.StandardContext.listenerStart Exception sending context initialized event to listener instance of class [com.helger.phase4.peppol.server.servlet.Phase4PeppolWebAppListener]
	com.helger.commons.exception.InitializationException: Failed to load configured private key
		at com.helger.phase4.peppol.server.servlet.Phase4PeppolWebAppListener._initPeppolAS4(Phase4PeppolWebAppListener.java:203)
		at com.helger.phase4.peppol.server.servlet.Phase4PeppolWebAppListener.initManagers(Phase4PeppolWebAppListener.java:238)
		at com.helger.photon.core.servlet.WebAppListener.contextInitialized(WebAppListener.java:688)
		at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4768)
		at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5230)
```

**Comment by michielbdejong** [comment]
ok, next one:
```
The provided certificate is not a valid Peppol certificate. Check result: UNSUPPORTED_ISSUER
```

**Comment by michielbdejong** [comment]
working on this in https://github.com/phax/phase4/issues/56

**Comment by michielbdejong** [comment]
OK, so as answered there, the problem is that `phase4-peppol-server-webapp` is Peppol-specific, and the certs pondersource-c2 and pondersource-c3 certs we received from for "eDelivery AS4" testing, are not.

The certs seem to be its own issuer (does that mean it's self-signed?):
```sh
root@c2:/usr/local/tomcat# keytool -v -list -keystore /as4Keys/as4.pfx -alias pondersource-c2
Enter keystore password:  
Alias name: pondersource-c2
Creation date: Jun 30, 2021
Entry type: PrivateKeyEntry
Certificate chain length: 1
Certificate[1]:
Owner: CN=pondersource-c2, OU=pondersource-c2, O=pondersource-c2, L=pondersource-c2, ST=pondersource-c2, C=EU
Issuer: CN=pondersource-c2, OU=pondersource-c2, O=pondersource-c2, L=pondersource-c2, ST=pondersource-c2, C=EU
Serial number: b824c0
Valid from: Wed Jun 30 15:24:00 UTC 2021 until: Fri Jun 30 15:24:00 UTC 2023
Certificate fingerprints:
	 SHA1: 38:26:D9:9F:00:2E:08:ED:AA:EF:46:3E:D7:83:76:98:7E:83:D6:76
	 SHA256: 11:3F:3F:08:10:30:AA:46:BD:64:DF:6D:8B:1D:F7:99:50:36:07:D6:77:A8:6D:6A:E1:B5:7B:50:4D:C5:41:18
Signature algorithm name: SHA256withRSA
Subject Public Key Algorithm: 2048-bit RSA key
Version: 3

Extensions: 

#1: ObjectId: 2.5.29.14 Criticality=false
SubjectKeyIdentifier [
KeyIdentifier [
0000: 79 89 12 EA 15 13 DE 6F   1C 13 ED F5 37 7D 63 7D  y......o....7.c.
0010: 7E B3 5A B8                                        ..Z.
]
]

```
So I can think of a few ways to fix it:
* pretend that whatever the issuer is, is really in the list of trusted Peppol issuers
* remove the issuer check from the code and re-compile
* build a non-Peppol web app

I think the second is easiest. There will not be any checks in this testbed that test whether a non-Peppol-issued cert is rejected.

**Comment by michielbdejong** [comment]
Seems that it's now passing the issuer check. I should update that to use an env var, since skipping that check is obviously security-sensitive.

It's now failing to load the private key: https://github.com/phax/ph-commons/blob/415b86e9509a4125132d59c6d86014c46977c7a2/ph-security/src/main/java/com/helger/security/keystore/KeyStoreHelper.java#L266

I wonder how I can access/edit the source of phax/ph-commons. Maybe in the build container?

**Comment by michielbdejong** [comment]
Ah, after trying many things, trying to build simple projects with maven and IntelliJ that somehow require ph-security, I found  I could add a file to ph-commons and compile that. now i should just find out how to run it, and then i can debug the private key loading error...

**Comment by michielbdejong** [comment]
```
java -cp ph-security/target/classes com.helger.security.keystore.App
Error: Unable to initialize main class com.helger.security.keystore.App
Caused by: java.lang.NoClassDefFoundError: com/helger/commons/id/IHasID

**Comment by michielbdejong** [comment]
hm, same problem with maven-assembly-plugin:

```
mvn compile package -DskipTests
java -cp ph-security/target/ph-security-10.1.3-SNAPSHOT.jar com.helger.security.keystore.App
Error: Unable to initialize main class com.helger.security.keystore.App
Caused by: java.lang.NoClassDefFoundError: com/helger/commons/id/IHasID

**Comment by michielbdejong** [comment]
Got the code compiling and on vps the incremental compile time is about 15 seconds so that's doable. Now I can debug. :)

**Comment by michielbdejong** [comment]
Ah, turns out that storepass should be a string but keypass should be a char array. Fixed that, and now can run ph-security test app to load the key from the .pfx.
Turns out that the problem in phase4 instance was a different one, I had corrected /root/run-c2.sh on cs.pondersource.net but not regenerated /root/phase4.properties so it was still looking for key name 'as4' instead of keyname 'pondersource-c2'.

That all fixed now, new error is:
```sh
root@c2:/usr/local/tomcat# cat logs/localhost.2021-07-14.log 
14-Jul-2021 10:22:46.611 SEVERE [main] org.apache.catalina.core.StandardContext.listenerStart Exception sending context initialized event to listener instance of class [com.helger.phase4.peppol.server.servlet.Phase4PeppolWebAppListener]
	com.helger.commons.exception.InitializationException: Failed to load configured Keystore
		at com.helger.phase4.peppol.server.servlet.Phase4PeppolWebAppListener._initPeppolAS4(Phase4PeppolWebAppListener.java:198)
		at com.helger.phase4.peppol.server.servlet.Phase4PeppolWebAppListener.initManagers(Phase4PeppolWebAppListener.java:238)
		at com.helger.photon.core.servlet.WebAppListener.contextInitialized(WebAppListener.java:688)
		at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4768)
		at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5230)
		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
		at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:728)
		at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:700)
		at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:696)
		at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:1024)
		at org.apache.catalina.startup.HostConfig$DeployWar.run(HostConfig.java:1911)
		at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)

**Comment by michielbdejong** [comment]
I'll see if I can add some debug statements into my branch of phase4 and then rebuild the Docker image so I get some more output there.

**Comment by michielbdejong** [comment]
root@c2:~/phase4# find . -name Phase4PeppolWebAppListener.java
./phase4-peppol-server-webapp/src/main/java/com/helger/phase4/peppol/server/servlet/Phase4PeppolWebAppListener.java

**Comment by michielbdejong** [comment]
I asked CEF to configure the test endpoints as https://c2.pondersource.net/as4/ and https://c3.pondersource.net/as4/
But the first test I tried to run in 
https://edelconf.westeurope.cloudapp.azure.com:9001/getTestSuiteDetailView?id=191&display=status returned that it was expecting a SOAP response and not a text/html response on there.
That makes sense, but then I'll first have to read up on how the c1-c2 and c2-c4 communication works for the [AS4 Basic Connectivity Tests](https://ec.europa.eu/cefdigital/wiki/display/CEFDIGITAL/eDelivery+Connectivity+testing).

**Comment by michielbdejong** [comment]
That was fixed in https://github.com/michielbdejong/phase4/commit/4225aa6d8e49955207c0ddc4ad656bddf1541c97.

I'm not more in debugging the testbed mode.

I received their instructions, which are not really public, although the license says they are, but they are available from them for free and they prefer if you get them from them, so they can track their readers. They were in docx format and not really readable on Mac but I imported them to this non-public Google Drive URL:https://docs.google.com/document/d/1RMM7iaILrLBXLE_Ksjl6WoohE4l5X45o/

So we now have a working image for Phase4, I'll close this issue and define two follow-ups:

* #6 
* #7
